### ISSUE TITLE
[RFC] explicit variable scope / not annotated uninitialized

### ISSUE THREAD

**Issue by bcardiff** [crystal-lang/crystal#3215]
Sometimes I find myself in the following pattern

```crystal
obj = uninitialized Foo
some_with_that_exec_block do |context|
   # ...
   obj = context.foo
end
do_something_with obj
```

The code is unsafe, for sure. And I don't want/need to change that. 
What bugs me is the mention of the `Foo`.

Crystal allows you to write

```crystal
some_with_that_exec_block do |context|
   # ...
   obj = context.foo
  do_something_with obj
end
```

but in order to declare that the scope of `obj` is outside the `do |context|` block, then a type annotation is needed, yet the compiler is able to know the type of `obj`

**Proposal 1**

Allow `uninitialized` to be used without type annotation. It keeps the unsafe nature of uninitialized.

```crystal
obj = uninitialized
some_with_that_exec_block do |context|
   # ...
   obj = context.foo
end
do_something_with obj
```

**Proposal 2**

Allow `_` or maybe any other type restrictions over `uninitialized`. It will not compile if the types are not assignable. It is a generalization of current syntax.

```crystal
obj = uninitialized _
some_with_that_exec_block do |context|
   # ...
   obj = context.foo
end
do_something_with obj
```

**Proposal 3**

Allow just `_` as a declaration of the scope of the variable. I don't like this very much because it hides the uninitialized value of the variable. It adds an overload to `_`. But I imagine is something someone will propose. Seems less syntax, but more cryptic IMO.

```crystal
obj = _
some_with_that_exec_block do |context|
   # ...
   obj = context.foo
end
do_something_with obj
```

**Comment by jhass** [comment]
First of all let me mention another (safer) variant, how you would solve it in Ruby:

```cr
obj = nil
some_with_that_exec_block do |context|
   # ...
   obj = context.foo
end
do_something_with obj.not_nil!
```

similarly extracting into a method works:

```cr
def fetch_foo
  some_with_that_exec_block do |context|
    # ...
    return context.foo
  end
  raise
end
do_something_with fetch_foo
```

Ruby actually has a feature for the inverse, preventing a closure:

```rb
foo = 1
bar do |context; foo|
  foo = 2
end
foo # => 1
```

I think we should find something along those lines, if we want to do this at all. Perhaps
```cr
some_with_that_exec_block do |context; obj=|
   # ...
   obj = context.foo
end
do_something_with obj
```

**Comment by bcardiff** [comment]
@jhass I like the prevent closure `;`. It will end up in nillable values to be safe for sure.

Why in the last example there is a `obj=` rather than a `obj`?

**Comment by jhass** [comment]
Preventing a closure is not doing what you want, on the contrary, so I'm trying to come up with something for the inverse, pushing a block local to the parent scope.

**Comment by bcardiff** [comment]
@jhass yes. I meant preventing making the variable local to the clousure. Pushing up to the parent scope is far more clear way to express it. 

I still think the `=` is a typo, isn't it?

Unless there is other use case for the not annotated `uninitialized` it's fine to make a RFC towards `;`. It seems more neat.

**Comment by jhass** [comment]
A closure means the nested scope closures over the outer scope, the variable declared in the outer scope remains to be the same variable in the inner scope. Thus preventing a closure means introducing a new local variable in the nested scope. This is not what you're looking for, but it's what `;` does in Ruby. Have a second look at my above example, here's a pry session demonstrating it:
```rb
[18] pry(main)> def foo
[18] pry(main)*   yield
[18] pry(main)* end
=> :foo
[19] pry(main)> x = 1; y = 1
=> 1
[20] pry(main)> foo do |; x|
[20] pry(main)*   x = 2
[20] pry(main)*   y = 2
[20] pry(main)* end
=> 2
[21] pry(main)> x
=> 1
[22] pry(main)> y
=> 2
```

So, introducing it like that with actually the *opposite behavior* would be rather confusing. Yet the overall area of the two features is the same, so I was searching for something that resembles it.

**Comment by jhass** [comment]
#1395 could be extended to allow this too actually

```cr
some_with_that_exec_block do |context|
   # ...
   $obj = context.foo
end
do_something_with $obj

**Comment by ozra** [comment]
When doing monkey business, I think it's perfectly good to be a little explicit; adding the type above seems a perfectly ok price for getting away with a little black magic. Chances are though, that LLVM would compile the `obj = nil`... version (which no doubt is cleaner) to the same thing after optimizations, if verified, that would be the best route.

Personally, I don't like the "generalized magic vars" theme, I'd rather see them gone completely.

The `do |a, b; x, y]` thing would be lovely for explicitly saying what _is_ closured imo. Preferably with ref as default, and copied as optional ("capturing" the value), via say `let`/`var` / whatever. (`do |a, b; let x, var y`) - would be handy for some async work. I can see how that can't happen given the Ruby-situation though.

**Comment by RX14** [comment]
I very much prefer the `obj = nil` version, because in the case where the block isn't executed or for some reason the variable *isn't* set, you get an exception instead of a segfault/other undefined behaviour.

**Comment by straight-shoota** [comment]
I actually like the idea of using a separate (private) method for this (https://github.com/crystal-lang/crystal/issues/3215#issuecomment-243545918). It's safe because it raises if no `Foo` could be found, but avoids a nilable union type. It's obviously not going to be an easy solution if you need closured variables in the block, because it doesn't share the scope. But if not, this is pretty perfect solution. A private method shouldn't add much overhead.

Continuing on this, maybe there could be a way to recreate the behaviour which can currently only be implemented using a separate method, to be moved in scope? But I suppose a method is actually the best way to model the control flow.
